lib/openstack_query/query_parser.py

```diff
@@ -0,0 +1,211 @@
+            ),
+        )
+
+    def parse_group_by(
+        self,
+        group_by: PropEnum,
+        group_ranges: Optional[Dict[str, List[PropValue]]] = None,
+        include_missing: Optional[bool] = False,
+    ) -> None:
+        """
+        Public method used to configure grouping results.
+        :param group_by: name of the property to group by
+        :param group_ranges: a dictionary containing names of the group and list of prop values
+        to select for that group
+        :param include_missing: a flag which if set includes an extra grouping for values
+        that don't fall into any group specified in group_ranges
+        """
+        self._check_prop_valid(group_by)
+        self._group_by = group_by
+        all_prop_list = set()
+
+        if group_ranges:
+            logger.debug("creating filter functions for specified group ranges")
+            for name, prop_list in group_ranges.items():
+                group_vals = tuple(prop_list)
+                self._group_mappings.update(
+                    {
+                        name: (
+                            lambda obj, lst=group_vals: self._prop_handler.get_prop(
+                                obj, group_by
+                            )
+                            in lst
+                        )
+                    }
+                )
+                all_prop_list.update(prop_list)
+
+            # if ungrouped group wanted - filter for all not in any range specified
+            if include_missing:
+                logger.debug("creating filter function for ungrouped group")
+                self._group_mappings.update(
+                    {
+                        "ungrouped results": lambda obj: self._prop_handler.get_prop(
+                            obj, group_by
+                        )
+                        not in all_prop_list
+                    }
+                )
+
+    def run_parser(
+        self, obj_list: List[OpenstackResourceObj]
+    ) -> Union[List[OpenstackResourceObj], Dict[str, List[OpenstackResourceObj]]]:
+        """
+        Public method used to parse query runner output - performs specified sorting and grouping
+        :param obj_list: a list of openstack resource objects to parse
+        """
+        if not obj_list:
+            return obj_list
+
+        # we sort first - assuming sorting is commutative to grouping
+        if self._sort_by:
+            obj_list = self._run_sort(obj_list, self._sort_by)
+
+        if self._group_by:
+            # if group mappings not specified - make a group for each unique value found for prop
+            if not self._group_mappings:
+                logger.info(
+                    "no group ranges specified - grouping by unique values of %s property",
+                    self._group_by.name,
+                )
+                self._group_mappings = self._build_unique_val_groups(
+                    obj_list, group_by_prop=self._group_by
+                )
```

Comment:
Nit: The `group_by` method should do this, not `run_parser`

Reply:
neither `parse_group_by` or `QueryMethods.group_by` methods can build unique value groups as the query needs to be run first. I can move this out into `run_group_by` method

Reply:
Yeah I'm thrown by us having the methods which prepare the grouping before the query, and the methods that actually do the grouping after in the same file

The ones for prepping are fine IMO (since it's a parser) but the ones that actively group should be elsewhere. But that's just my 2 cents